src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ReflectMethods.java

```diff
@@ -1494,358 +1494,251 @@ public void visitIf(JCTree.JCIf tree) {
-                    bodies.add(stack.body);
+                    localResult = append(ExtendedOp.conditionalOr(clBodies));
+                }
 
-                    // Pop labels
-                    popBody();
-                } else if (headCl instanceof JCTree.JCDefaultCaseLabel) {
-                    // @@@ Do we need to model the default label body?
-                    pushBody(headCl, FunctionType.VOID);
+                append(CoreOp._yield(localResult));
+                body = stack.body;
 
-                    append(CoreOp._yield());
-                    defaultLabel = stack.body;
+                // Pop labels
+                popBody();
+            } else if (headCl instanceof JCTree.JCDefaultCaseLabel) {
+                // @@@ Do we need to model the default label body?
+                pushBody(headCl, FunctionType.VOID);
 
-                    // Pop label
-                    popBody();
-                } else {
-                    throw unsupported(tree);
-                }
+                append(CoreOp._yield());
+                body = stack.body;
 
-                // Statements body
-                switch (c.caseKind) {
-                    case RULE -> {
-                        pushBody(c.body, actionType);
-                        if (c.body instanceof JCTree.JCBlock b) {
-                            toValue(b);
-                            if (!(b.stats.last() instanceof JCTree.JCBreak)) {
-                                append(CoreOp._yield()); // @@@ _break is also an option
-                            }
-                        }
-                        else if (c.body instanceof JCTree.JCStatement s) {
-                            toValue(s);
-                            if (!(s instanceof JCTree.JCThrow)) {
-                                append(CoreOp._yield());
-                            }
-                        }
+                // Pop label
+                popBody();
+            } else {
+                throw unsupported(tree);
+            }
 
-                        if (headCl instanceof JCTree.JCDefaultCaseLabel) {
-                            defaultStatements = stack.body;
-                        } else {
-                            bodies.add(stack.body);
-                        }
+            return body;
+        }
 
-                        // Pop block
-                        popBody();
+        private Body.Builder visitCaseBody(JCTree tree, JCTree.JCCase c) {
+            Body.Builder body = null;
+
+            FunctionType caseBodyType;
+            Type yieldType = null;
+            if (tree instanceof JCTree.JCSwitch) {
+                caseBodyType = FunctionType.VOID;
+            } else if (tree instanceof JCTree.JCSwitchExpression) {
+                Type switchType = adaptBottom(tree.type);
+                caseBodyType = FunctionType.functionType(typeToTypeElement(switchType));
+                yieldType = adaptBottom(tree.type);
```

Comment:
Are we not already passing the FunctionType to visitSwitchStatAndExpr? So the only difference would be the yieldType.